Server/network/raknet/RakNetHandler.hpp

```diff
@@ -54,154 +43,45 @@ class RakNetHandler {
-			stream->size = packet->bitSize;
-			bool compressionEnabled = false;
-			BinaryStream* streams = new BinaryStream(NULL);
-			streams->size = 0;
-			size_t streams_count = 0;
+	    stream->size = packet->bitSize;
+            bool compressionEnabled = false;
+	    BinaryStream* streams = new BinaryStream(NULL);
+	    streams->size = 0;
+	    size_t streams_count = 0;
             switch(stream->get_unsigned_byte()) {
                 // Minecraft packet
                 case 0xFE: {
                     Logger::log("MCPE Game packet recieved.");
-					stream->put_unsigned_byte(0xFE);
-					stream->offset += stream->size;
-					if (!compressionEnabled) {
-						BinaryStream* data = new BinaryStream(NULL);
-						data->buffer = (int8_t *)(Bytef*)(stream->buffer + 1);
-						data->size = (uLong)(static_cast<size_t>(stream->size) - 1);
-						data->offset = 0;
-						do {
-							BinaryStream* entry_stream = new BinaryStream(NULL);
-							entry_stream->size = data->get_var_int();
-							entry_stream->buffer = data->get_bytes(entry_stream->size);
-							entry_stream->offset = 0;
-							++streams_count;
-							streams->buffer = entry_stream->buffer;
-							streams = (BinaryStream*)realloc(streams, streams_count * sizeof(BinaryStream));
-							if (streams->get_var_int() == 0xc1) { // request network settings
-								std::cout << "Player protocol version: " << streams->get_int_be();
-								std::cout << std::endl;
-								compressionEnabled = true;
-								free(streams->buffer);
-							}
-						} while(data->offset < data->size);
-					}
-					else {
-						std::cout << "Compression reach" << std::endl;
-						zlib_buf_t in;
-						in.data = (Bytef*)(stream->buffer + 1);
-						in.size = (uLong)(stream->size - 1);
-						zlib_buf_t out;
-						zlib_decode(in, &out, ZLIB_RAW_MODE);
-						BinaryStream* data = new BinaryStream(NULL);
-						data->buffer = (int8_t*)out.data;
-						data->size = out.size;
-						data->offset = 0;
-						free(data->buffer);
-					}
+		    stream->put_unsigned_byte(0xFE);
+		    stream->offset += stream->size;
+		    if (!compressionEnabled) {
+			   BinaryStream* data = new BinaryStream(NULL);
+			   data->buffer = (int8_t *)(Bytef*)(stream->buffer + 1);
+			   data->size = (uLong)(static_cast<size_t>(stream->size) - 1);
+			   data->offset = 0;
+			   do {
+				   BinaryStream* entry_stream = new BinaryStream(NULL);
+				   entry_stream->size = data->get_var_int();
+				   entry_stream->buffer = data->get_bytes(entry_stream->size);
+				   entry_stream->offset = 0;
+				   ++streams_count;
+				   streams->buffer = entry_stream->buffer;
+				   streams = (BinaryStream*)realloc(streams, streams_count * sizeof(BinaryStream));
+				   if (streams->get_var_int() == 0xc1) { // request network settings
+					   std::cout << "Player protocol version: " << streams->get_int_be();
+					   std::cout << std::endl;
+					   compressionEnabled = true;
+					   free(streams->buffer);
```

Comment:
It looks like you are only freeing uint8_t*.  You can create a destructor for the BinaryStream which will destroy the BinaryStream and all the data in it.
![__](https://user-images.githubusercontent.com/60012371/212541329-d6313e8f-dc04-4151-85ac-e36d9b66c146.png)

Reply:
You can change this to `free((void *) streams->buffer);`